pkg/database/database.go

```diff
@@ -50,33 +68,174 @@ func updateFromEnv(dsn string) (string, error) {
 	return result, nil
 }
 
-func Connect(ctx context.Context, dsn string, user string, password string) (*sql.DB, DbEngine, error) {
-	populatedDSN, err := updateFromEnv(dsn)
+// extractPlaceholders replaces ${...} placeholders with unique numeric sentinels
+// and looks up the environment variable values immediately.
+// Numeric sentinels (999000, 999001, etc.) are valid in all URL components:
+// ports (must be numeric), hostnames, userinfo, paths, and query strings.
+// This allows us to parse the URL structure before expanding environment variables.
+// Returns: the string with sentinels, a mapping of sentinel->value, and any error.
+func extractPlaceholders(s string) (string, map[string]string, error) {
+	mapping := make(map[string]string)
+	counter := 0
+	var err error
+
+	result := DSNREnvRegex.ReplaceAllStringFunc(s, func(match string) string {
+		sentinel := fmt.Sprintf("999%03d", counter)
```

Comment:
Won't this replace any value that contains three 9s followed by three more numbers? eg: If DB_PASSWORD is "1234" and the connection string is `mysql://admin:${DB_PASSWORD}@localhost:3306/db-999000`, then the final DSN will be `mysql://admin:1234@localhost:3306/db-1234`

Reply:
You can also run into this if you put certain values in environment variables. eg: 

```golang
env: map[string]string{
	"DB_USER":     "user-999001",
	"DB_PASSWORD": "1234",
},
dsn:  "mysql://${DB_USER}:${DB_PASSWORD}@localhost:3306/dbname",
```

...will generate `mysql://user-1234:1234@localhost:3306/dbname` instead of `mysql://user-999001:1234@localhost:3306/dbname`